### PR TITLE
implement rule-based sampler

### PR DESCRIPTION
**Pull request by brettmc**
Ported rule-based sampler from https://github.com/Nevay/otel-sdk-contrib-sampler

Co-authored-by: Tobias Bachert <git@b-privat.de>

### CODE REVIEW

**Comment by brettmc** [comment]
I think this is ready for review now, @open-telemetry/php-approvers 
There's an [OTEP](https://github.com/open-telemetry/oteps/pull/250) which looks like it will eventually specify a rules-based sampler, so to avoid a future config collision I've changed the key for this one to `contrib_rule_based` (open to improved names, though!)

File: src/Sampler/RuleBased/src/SamplingRule.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    public function matches(
+        ContextInterface $context,
+        string $traceId,
+        string $spanName,
+        int $spanKind,
+        AttributesInterface $attributes,
+        array $links,
+    ): bool;
+
+    public function __toString(): string;
```
[the remark this answers is not in the document]
Comment (brettmc) [reply src/Sampler/RuleBased/src/SamplingRule.php]: done.

File: src/Sampler/RuleBased/src/RuleSetInterface.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+use OpenTelemetry\SDK\Trace\SamplerInterface;
+
+interface RuleSetInterface
+{
+    /**
+     * @return list<SamplingRule>
+     */
+    public function samplingRules(): array;
+    public function delegate(): SamplerInterface;
+    public function __toString();
```
Comment (ChrisLightfootWild) [inline src/Sampler/RuleBased/src/RuleSetInterface.php:16]: Another `Stringable` maybe here?

**Review by ChrisLightfootWild — APPROVED** [review]
Apologies for the delay :snail:

I've been reading this as a reference point for the ComponentProvider configuration.

Maybe we could add `.gitattributes` as well to reduce package size, but happy for that to be a follow up.

**Pull request merged**